### DISCUSSION TITLE
Make it possible to re-order `LineItem`s

### DISCUSSION THREAD

**Discussion by freder** [craftcms/commerce#3129] [Ideas]
I need to sort the items in a cart by certain criteria.

I can do that in an adjuster, in which case the response of `/actions/commerce/cart/update-cart` has to correct order (because it returns the in-memory `Order` object). however, I also need to _persist_ the custom order, so that the backend, invoice pdf, etc. also use the correct order!

setting `$lineItem->dateCreated = /* ... */;` does not work because [it never actually gets stored in the db again](https://github.com/craftcms/commerce/blob/5c2f35ea4134f7d234f6e753fb46df0d986c3e0a/src/services/LineItems.php#L233-L259). — the only way to currently do it is to re-implement (and maintain) a slightly different version of [saveLineItem()](https://github.com/craftcms/commerce/blob/5c2f35ea4134f7d234f6e753fb46df0d986c3e0a/src/services/LineItems.php#L205) 😞 

---

PR: https://github.com/craftcms/commerce/pull/3130

**Comment by brandonkelly** [comment]
This would be against existing conventions. For example, `craft\services\Elements::saveElement()` will only ever set the element record’s `dateCreated` value for brand new elements:

https://github.com/craftcms/cms/blob/f4aafeece99bab6c1812e34c13612b128ddf486d/src/services/Elements.php#L3013-L3016

So I’d vote against this change.

**Reply by freder** [reply]
unless you change the value explicitly it would keep it's original value though.
a user wouldn't be able to change it, but a plugin dev _could_ if needed.

alternatively, if changing `dateCreated` is too hacky, you could provide an official/explicit way to re-order line items.

**Reply by freder** [reply]
the use case makes sense to you though, right?

**Comment by nfourtythree** [comment]
This is certainly a good feature request and the ability to have proper control over the order of line items is definitely something we will look into in the future. We will keep this FR open.

In the meantime, you are able to tweak the `dateCreated` on the `LineItem` record using the `EVENT_BEFORE_INSERT` and `EVENT_BEFORE_UPDATE` on the active record class.

Here is an example of setting every new insert with the date `1970-01-01 01:59:59`:

```php
use craft\commerce\records\LineItem;
use yii\base\ModelEvent;

// ...

Event::on(
    LineItem::class,
    LineItem::EVENT_BEFORE_INSERT,
    function(ModelEvent $event) {
        $event->sender->dateCreated = Db::prepareDateForDb(new \DateTime('1970-01-01 01:59:59'));
    }
);
```

Hope this helps in the short term, thanks!

**Reply by freder** [reply]
that _should_ work, yes.

---

what I have so far is a bit simpler in that it doesn't require me to figure out the exact date that would insert the line item at the correct position:

```php
Event::on(
	Order::class,
	Order::EVENT_BEFORE_SAVE,
	function (ModelEvent $event) {
		$order = $event->sender;

		// [...] 
		
		// sort the line items
		$lineItemsOrdered = sortCart($order);

		$idx = 0;
		foreach ($lineItemsOrdered as $item) {
			$item->dateCreated = new \DateTime('+' . $idx . ' seconds');
			$item->dateUpdated = $item->dateCreated;
			$idx += 1;
		}
		$order->setLineItems($lineItemsOrdered);

		// [...]

		foreach ($order->getLineItems() as $item) {
			// modified version of https://github.com/craftcms/commerce/blob/5c2f35ea4134f7d234f6e753fb46df0d986c3e0a/src/services/LineItems.php#L205
			saveLineItem($order, $item, false);
		}
	}
);
```
